### PR TITLE
Mirroring script: Edge mirrors from both IE AND Chrome at the same time

### PR DESCRIPTION
**Pull request by queengooborg**
This PR refactors the mirroring script, adding new functions (and a few bug fixes), to mirror data for Edge a little bit differently in light of its engine change.  Now, rather than mirroring from just IE or just Chrome, it will mirror the data from _both_ browsers at once.

This includes the addition of a `combineStatements()` function, which takes an array of support statements and combines them down to their simplest form possible.  This will be helpful for future multi-source mirroring to come.  Accordingly, this also provides a structure for multi-source mirroring.

Inspired by #6575.

### CODE REVIEW

File: docs/contributing.md [hunk 1]
```diff
@@ -74,18 +74,18 @@ If the feature you're interested in is a JavaScript API, you can cross-reference
...
-| ---------------- | ----------------- |
-| Chrome Android   | Chrome            |
-| Edge             | Internet Explorer |
-| Firefox Android  | Firefox           |
-| Opera            | Chrome            |
-| Opera Android    | Chrome Android    |
-| Safari iOS       | Safari            |
-| Samsung Internet | Chrome Android    |
-| WebView          | Chrome Android    |
+The <browser> argument is the destination browser that values will be copied to. The script automatically determines what browser to copy from based upon the destination (see table below), but manual specification is possible through the `--source=""` argument. (Mirroring data for Edge ignores this argument.)
```
Comment (foolip) [inline docs/contributing.md:77]: Where is the "Mirroring data for Edge ignores this argument" bit in the implementation? I can't spot it, but in any case I don't think that's the behavior we should have. Let's have `--source` override the special mirroring logic for Edge if it's used. (If it's a no-op, it would be better to have the command fail with an error message.)

File: scripts/mirror.js [hunk 2]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-    if (sourceData.notes) {
-      newData.notes = updateNotes(
-        sourceData.notes,
-        /Internet Explorer/g,
-        'Edge',
-      );
-    }
-  } else if (source == 'chrome') {
-    newData = originalData == undefined ? sourceData : originalData;
+  if (ieData.version_removed && ieData.version_removed !== null) {
```
Comment (foolip) [inline scripts/mirror.js:231]: There's a bug here. Running `npm run mirror -- --m always edge` throws "TypeError: Cannot read property 'version_removed' of undefined" on this line. Seems like it's failing in webextensions/api/alarms.json where there is no IE data.

**Comment by foolip** [comment]
I tried running `npm run mirror -- --m always edge` (with workaround for a bug in `bumpEdge` added) and then `git diff | grep '^+ ' | sed 's/^+ *//;s/,$//' | sort | uniq -c | sort -nr`  to see what kinds of changes this is mostly making.

It looks pretty reasonable, but it makes me suspect there are fixes that would improve this a lot. Most of the changes are adding "≤79", and very few "79" are added. There should be *lots* of APIs the data says was not in IE but was in Chrome ≤79, and where the value should be set to "79". The first cases in the diff of this is AbortController.json. Can you experiment with checking for false in the IE data, if that's not already being done?

Then there are lots of nulls being added. Sometimes that makes sense, if it was added in Edge 12-18 then obviously we can't infer the data from IE+Chrome. There are so many nulls that I suspect that there is still some bug here, but I've failed to find an incorrect change by skimming through the diffs. I can take another look after the PR has been updated.

File: scripts/mirror.js [hunk 3]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-const bumpEdge = (originalData, sourceData, source) => {
-  let newData = {};
-
-  if (source == 'ie') {
-    if (sourceData.version_removed && sourceData.version_removed !== null) {
-      newData.version_added = false;
-    } else if (sourceData.version_added !== null) {
-      newData.version_added = sourceData.version_added ? '12' : null;
-    }
+const bumpEdge = comp => {
```
Comment (foolip) [inline scripts/mirror.js:225]: Can you call this `data` instead? The compat statement objects are mostly called that in this file, and `comp` is more easily misread is being a comparator, like for a sort. (Not a misunderstanding that would stick for long, of course.)

File: scripts/mirror.js [hunk 4]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+  let newData = comp['edge'];
```
Comment (foolip) [inline scripts/mirror.js:226]: Assigning the same object to `newData` and `originalData` can't be right. Should `newData` be an empty object, or a copy of `comp['edge']`?

File: scripts/mirror.js (hunk 2 above)
Comment (foolip) [inline scripts/mirror.js:231]:
```suggestion
  if (ieData.version_removed !== null) {
```

This is an equivalent check. But it still looks a bit funny. Should it be `ieData.version_added && !ie_data.removed_removed`?

File: scripts/mirror.js [hunk 5]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+    newData.version_added = ieData.version_added ? '12' : null;
+  }
 
-    let chromeFalse =
-      sourceData.version_added === false ||
-      sourceData.version_removed !== undefined;
-    let chromeNull = sourceData.version_added === null;
+  let chromeFalse =
+    chromeData.version_added === false ||
+    chromeData.version_removed !== undefined;
```
Comment (foolip) [inline scripts/mirror.js:239]: This bit makes me wonder what happens for a feature behind a flag. This script doesn't seem to handle flags at all. The data for Edge seems to mostly have flags copied, but it seems to me there's still a but waiting here: if the feature was added behind a flag in Chrome, `chromeFalse` will still end up false here, and we'll mark it as supported in Edge. I've verified this by modifying Bluetooth.json locally.

I think we'll need a step where support statements are pre-processed to apply flags, alternative_name, and perhaps more.

It's fine to do that in a follow-up though, it's a generic problem, not specific to mirroring Edge.

File: scripts/mirror.js [hunk 6]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-            if (originalData.version_added == false) {
-              newData.version_added = '79';
-            } else if (originalData.version_added == null) {
-              newData.version_added = '≤79';
-            }
-          } else {
-            newData.version_added == sourceData.version_added;
+  if (originalData === undefined) {
+    newData.version_added = chromeFalse ? false : chromeNull ? null : '≤79';
+  } else {
```
Comment (foolip) [inline scripts/mirror.js:244]: You can turn this into an `else if` and save some indentation in the following.

File: scripts/mirror.js [hunk 7]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-            Number(sourceData.version_added) <= 79
-          ) {
-            if (originalData.version_added == false) {
-              newData.version_added = '79';
-            } else if (originalData.version_added == null) {
-              newData.version_added = '≤79';
-            }
-          } else {
-            newData.version_added == sourceData.version_added;
+  if (originalData === undefined) {
```
Comment (foolip) [inline scripts/mirror.js:242]: In my testing `originalData === undefined` never happens. It was assigned to an object above...

File: scripts/mirror.js [hunk 8]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+    newData.version_added = chromeFalse ? false : chromeNull ? null : '≤79';
+  } else {
+    if (!chromeFalse && !chromeNull) {
+      if (originalData.version_added == true) {
```
Comment (foolip) [inline scripts/mirror.js:246]: Looking at the original data doesn't quite make sense together with `--m always`. Should this code not derive a statement from IE + Chrome, then use the command line argument to decide whether or not to update?

File: scripts/mirror.js [hunk 9]
```diff
@@ -219,71 +219,61 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+        newData.version_added = '≤18';
+      } else {
+        if (
+          chromeData.version_added == true ||
```
Comment (foolip) [inline scripts/mirror.js:250]: Increasingly, I think version_added true for Chrome means we don't know enough. What we *could* infer if Chrome is true and IE is false is really "≥79", but that's not an option. In the absence of that, I think we just have to set Edge to true as well?

File: scripts/mirror.js [hunk 10]
```diff
@@ -527,9 +514,6 @@ const bumpVersion = (data, destination, source, originalData) => {
       case 'chrome_android':
         bumpFunction = bumpChromeAndroid;
         break;
-      case 'edge':
```
Comment (foolip) [inline scripts/mirror.js:530]: This dispatch was kinda nice though. I think we can keep it by doing `bumpFunction = () => bumpEdge(extra, stuff, here)` and then avoid special-casing Edge below.

This is a bit it would be good to have @ddbeck check out, though.

**Review by foolip — COMMENTED** [review]
Sweet stuff! I have a lot of comments, but that's not because it ain't sweet :)

File: docs/contributing.md (hunk 1 above)
Comment (vinyldarkscratch) [reply docs/contributing.md:77]: Actually, thinking about it, maybe we should remove this argument entirely?  In all honesty, the sole purpose of this argument was to allow for mirroring in Edge from either IE or Chrome -- it's an argument that's not exactly well-supported and allows for mirroring across different engines.  We wouldn't want to encourage someone trying to mirror Internet Explorer's data onto Safari, for example.

File: scripts/mirror.js (hunk 5 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:241]: I think we should do this in a follow-up, considering that it's a problem that applies to all browsers.

File: scripts/mirror.js (hunk 10 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:530]: I'm happy to do this, but I'm thinking it would be better to follow this up in another PR, since it involves a bit more of a rewrite than I feel comfortable with in this PR.

Comment (foolip) [reply scripts/mirror.js:530]: Hmm, I see, it's because the argument for `bumpEdge` isn't available here, right?

File: scripts/mirror.js (hunk 5 above)
Comment (foolip) [reply scripts/mirror.js:241]: Yep! Can you file an issue for that?

File: docs/contributing.md (hunk 1 above)
Comment (foolip) [reply docs/contributing.md:77]: I've never used it, so I wouldn't mind. It does seem a little bit plausible that someone might want to mirror Opera to Opera Android instead of using Chrome Android, in case a feature was added (or removed) from both. I'll defer to @ddbeck on whether this should stay or not.

File: scripts/mirror.js (hunk 10 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:530]: I removed my earlier comment because I realized there actually __is__ a way to work it back in!  😅

File: scripts/mirror.js [hunk 11]
```diff
@@ -219,55 +219,54 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
 };
 
 /**
- * @param {SupportStatement} comp
+ * @param {SupportStatement} compData
```
Comment (foolip) [inline scripts/mirror.js:222]:
```suggestion
 * @param {SupportStatement} originalData
```

File: scripts/mirror.js (hunk 5 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:238]: Filed as #6625!

File: docs/contributing.md (hunk 1 above)
Comment (foolip) [reply docs/contributing.md:77]: Resolved by commit d6a5b11af19cb22f0726f49c34bcbd550a60deec.

**Comment by foolip** [comment]
@vinyldarkscratch is this ready for final review now? It would be good to first verify that the mirror script still produces exactly the same changes for all other browsers, with/without `--source` and with/without `--modify=always`. It would also be good to review a diff of what the mirroring script would do for Edge before and after these changes, to see that there aren't any large regressions.

**Comment by foolip** [comment]
@vinyldarkscratch can you ping me when you'd like review of this again?

**Comment by vinyldarkscratch** [comment]
This is all good to go now, yep!

**Review by alshamiri5 — APPROVED** [review]

**Comment by foolip** [comment]
Oops, I didn't notice I was pinged for review two weeks ago :)

**Comment by foolip** [comment]
I tried running `npm run mirror -- --modify=always edge` but some of the suggested changes don't seem right. For example, in api/AbortController.json, version_added is being set to false, even though it was in Chrome 66. The right answer if ignoring the existing data should be "79". (It not using `--modify=always` it isn't modified.)

@vinyldarkscratch can you tweak the script until the output when using `--modify=always` is reasonable when spot checking?

**Comment by vinyldarkscratch** [comment]
This is all ready to go; everything's been double-checked for functionality.  I would love to get this PR merged sooner than later, as it paves the way for two new PRs I have planned (one that fixes the WebView mirroring, another that has Opera Android account for data from both Opera Presto and Chrome Android).

File: scripts/mirror.js [hunk 12]
```diff
@@ -220,70 +218,59 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
- * @param {SupportStatement} sourceData
- * @param {string} source
+ * @param {SupportStatement} chromeData
+ * @param {SupportStatement} ieData
  * @returns {SupportStatement}
  */
-const bumpEdge = (originalData, sourceData, source) => {
-  let newData = {};
+const bumpEdge = (originalData, chromeData, ieData) => {
+  let newData = copyStatement(originalData);
```
Comment (foolip) [inline scripts/mirror.js:226]: There's a bug here, or in `copyStatement`, which I don't think is really new, but it didn't manifest before. If you run `npm run mirror -- edge --modify=bool`, this change will be produced:

```diff
--- a/css/selectors/file-selector-button.json
+++ b/css/selectors/file-selector-button.json
@@ -26,11 +26,24 @@
             ],
             "edge": [
               {
+                "0": {
+                  "version_added": false
+                },
+                "1": {
+                  "version_added": "79",
+                  "alternative_name": "::-webkit-file-upload-button"
+                },
                 "version_added": false
               },
               {
-                "version_added": "79",
-                "alternative_name": "::-webkit-file-upload-button"
+                "0": {
+                  "version_added": false
+                },
+                "1": {
+                  "version_added": "79",
+                  "alternative_name": "::-webkit-file-upload-button"
+                },
+                "version_added": "79"
               }
             ],
             "firefox": {
```

It looks like Chrome's array of support statements is being treated as a single support statement, with broken results. It seems to me this might also be happening for other uses of `copyStatement`, but for this PR it's important to at least not regress what `npm run mirror` does.

**Review by foolip — COMMENTED** [review]
Sorry I left this sitting for months! I've taken another look and tried it a bit myself, unfortunately found a bug in need of fixing.

File: scripts/mirror.js (hunk 12 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:230]: I don't know why it took me this long to finally get a fix out, especially one so easy.  Probably just needed a fresh pair of eyes...  Anyways, this is all fixed and ready to go!

File: scripts/mirror.js [hunk 13]
```diff
@@ -223,6 +223,10 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
  * @returns {SupportStatement}
  */
 const bumpEdge = (originalData, chromeData, ieData) => {
+  if (Array.isArray(originalData)) {
+    return originalData.map(d => bumpEdge(d));
```
Comment (foolip) [inline scripts/mirror.js:227]: What about the `chromeData` and `ieData` arguments in this case? If they're arrays, should the data be used? How about if they're not arrays? Just ignoring the data seems odd, although I don't remember now how any of this works :)

File: scripts/mirror.js [hunk 14]
```diff
@@ -220,70 +218,63 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-            newData.version_added == sourceData.version_added;
-          }
-        }
-      } else if (chromeFalse) {
-        if (originalData.version_added && !originalData.version_removed) {
-          newData.version_removed = '79';
-        }
+  let chromeFalse =
+    chromeData.version_added === false ||
+    chromeData.version_removed !== undefined;
```
Comment (foolip) [inline scripts/mirror.js:242]: Could this be `chromeData.version_removed || chromeData.version_added === false`, more like the `ieData` case above?

File: scripts/mirror.js [hunk 15]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-    }
+  let newData = copyStatement(originalData);
 
-    if (sourceData.notes) {
-      newData.notes = updateNotes(
-        sourceData.notes,
-        /Internet Explorer/g,
-        'Edge',
-      );
+  if (ieData) {
```
Comment (foolip) [inline scripts/mirror.js:232]: `ieData` can be an array here, verified by adding an assert. Can you explicitly handle that case, even if it's by bailing and doing nothing?

File: scripts/mirror.js [hunk 16]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-          } else {
-            newData.version_added == sourceData.version_added;
-          }
-        }
-      } else if (chromeFalse) {
-        if (originalData.version_added && !originalData.version_removed) {
-          newData.version_removed = '79';
-        }
+  let chromeFalse =
+    chromeData.version_removed || chromeData.version_added === false;
```
Comment (foolip) [inline scripts/mirror.js:241]: It should be possible that `chromeData` is an array here, but adding an assert I don't see it happening. Can it happen, and if so can you handle it?

File: scripts/mirror.js [hunk 17]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-      newData.notes = updateNotes(
-        sourceData.notes,
-        /Internet Explorer/g,
-        'Edge',
-      );
+  if (ieData) {
+    if (ieData.version_removed || ieData.version_added === false) {
+      newData.version_added = false;
+    } else if (ieData.version_added) {
+      newData.version_added = ieData.version_added ? '12' : null;
```
Comment (foolip) [inline scripts/mirror.js:236]:
```suggestion
      newData.version_added = '12';
```

`ieData.version_added` is always truthy here.

File: scripts/mirror.js [hunk 18]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
-        if (originalData.version_added && !originalData.version_removed) {
-          newData.version_removed = '79';
-        }
+  let chromeFalse =
+    chromeData.version_removed || chromeData.version_added === false;
+  let chromeNull = chromeData.version_added === null;
+
+  if (!chromeFalse && !chromeNull) {
+    if (originalData.version_added === true) {
+      newData.version_added = '≤18';
```
Comment (foolip) [inline scripts/mirror.js:246]: This is an existing condition, but it doesn't make sense to me. If the original data was true, and Chrome is neither false nor null, then I don't think anything at all follows, that depends on what the Chrome data is.

Can you find a case where this branch is taken and see if it actually makes sense there? I'm skeptical :)

File: scripts/mirror.js [hunk 19]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+    } else if (chromeData.version_added === true) {
+      newData.version_added = true;
+    } else if (Number(chromeData.version_added) <= 79) {
+      if (
+        originalData.version_added === false ||
+        newData.version_added === false
```
Comment (foolip) [inline scripts/mirror.js:252]: OK, so `newData.version_added === false` can happen here if it was set to false because it wasn't in IE. This makes sense, but took some pondering to understand :)

A comment for each case would be nice, but I won't insist on it.

File: scripts/mirror.js [hunk 20]
```diff
@@ -220,70 +218,62 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
       }
+    } else {
+      newData.version_added = chromeData.version_added;
     }
+  } else if (chromeFalse) {
+    if (originalData.version_added && !originalData.version_removed) {
+      newData.version_removed = '79';
+    }
+  } else if (chromeNull) {
+    newData.version_added = null;
```
Comment (foolip) [inline scripts/mirror.js:266]: There soon won't be any null Chrome values, but can't this case end up overwriting non-null data sourced from IE? Is it needed?

**Review by foolip — CHANGES_REQUESTED** [review]
This is pretty close to done I think, let's get it across the finish line!

File: scripts/mirror.js (hunk 15 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js]: I've just pushed a commit that includes something that solves this issue and beyond!

File: scripts/mirror.js (hunk 18 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:290]: This is for if Chrome is `true`, though the way the code is structured doesn't help with legibility.  Fixing that up!

File: scripts/mirror.js (hunk 20 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:310]: I've made some changes which will prevent this from occurring!

**Comment by vinyldarkscratch** [comment]
Got an improved structure for multi-source mirroring put into place!  I think this is all ready to go now!

File: scripts/mirror.js [hunk 21]
```diff
@@ -219,74 +280,93 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+      newData.push(bumpEdgeFromIE(ieData));
     }
   }
 
-  return newData;
+  if (chromeData) {
+    if (Array.isArray(chromeData)) {
+      newData += chromeData.map(d => bumpEdgeFromChrome(d, originalData));
+    } else {
+      newData.push(bumpEdgeFromChrome(chromeData, originalData));
```
Comment (foolip) [inline scripts/mirror.js:365]: If you run `npm run -- mirror edge -m always` an exception will be thrown here: "newData.push is not a function"

File: scripts/mirror.js [hunk 22]
```diff
@@ -219,74 +280,93 @@ const bumpChromeAndroid = (originalData, sourceData, source) => {
...
+    return originalData.map(d => bumpEdge(d, chromeData, ieData));
+  }
 
-    if (newNotes) {
-      newData.notes = newNotes;
+  let newData = [];
+
+  if (ieData) {
+    if (Array.isArray(ieData)) {
+      newData += ieData.map(d => bumpEdgeFromIE(d));
```
Comment (foolip) [inline scripts/mirror.js:355]: This and the the similar branch won't work, arrays can't be extended like this in JS.

**Review by foolip — COMMENTED** [review]
Found some bugs. It would be good to try running the script with all permutations of arguments and seeing if the result always makes sense, passes the lint, and whatever else can be easily checked.

File: scripts/mirror.js (hunk 22 above)
Comment (vinyldarkscratch) [reply scripts/mirror.js:355]: Ah, right, sometimes I forget Python and JavaScript are two separate languages.  :P

Comment (foolip) [reply scripts/mirror.js]: Me too :)

**Comment by foolip** [comment]
@vinyldarkscratch have you tried running without various arguments to see what kinds of changes are made? Can you summarize for ease of review?

**Comment by foolip** [comment]
@teoli2003 ran into the fact that Edge is mirrored only from IE in https://github.com/mdn/browser-compat-data/pull/11063#discussion_r653635715 so I have a renewed desire to get this landed :)

**Comment by vinyldarkscratch** [comment]
Running the mirroring script now without arguments (other than `edge` of course), I've noticed a few cases that may still need to be resolved.  Looking at the [generated diff](https://github.com/mdn/browser-compat-data/files/6735831/mirror.diff.txt), these include:

- IE is `false`, Chrome is `null` -> Edge becomes `false` (should be `null`)
- IE is truthy, Chrome has a `version_added` and `version_removed` -> Edge becomes multi-statement and Chrome's versions don't get bumped

**Comment by vinyldarkscratch** [comment]
Alright, this is now ready to go, I'm happy to say!  All the major bugs have been worked out, and I'm confident in the results yielded by this script now!

File: scripts/mirror.js [hunk 23]
```diff
@@ -187,13 +183,91 @@ const copyStatement = data => {
...
+  const ignored_keys = ['version_added', 'notes'];
+
+  let flattenedData = data.flat(2);
+  let sections = {};
+  let newData = [];
+
+  for (const d of flattenedData) {
+    let key = Object.keys(d)
+      .filter(k => !ignored_keys.includes(k))
+      .join('');
```
[the remark this answers is not in the document]
Comment (queengooborg) [reply scripts/mirror.js:200]: This `key` parameter here is to make sure that multiple statements with the same set of keys are properly merged, so this is fine!  Joining them together just makes it a little easier to work with.

**Comment by queengooborg** [comment]
I mentioned this during our last meeting -- although there are still some unresolved comments, I feel this is ready enough to merge now, and then any more changes can be followed up later.  This PR's blocking some other mirroring updates, and I think it's ready enough to use!

**Comment by foolip** [comment]
@queengooborg I just tried `npm run mirror -- -m bool edge` and that throws "TypeError: Invalid argument expected string". Can you check that out?

**Comment by queengooborg** [comment]
Slight logic problem, fixed now though!

**Comment by foolip** [comment]
`npm run mirror -- -m bool edge` now removes notes which it didn't do before. Example:
```diff
--- a/api/CSSStyleDeclaration.json
+++ b/api/CSSStyleDeclaration.json
@@ -163,8 +163,7 @@
               "notes": "See <a href='https://crbug.com/331608'>bug 331608</a>."
             },
             "edge": {
-              "version_added": false,
-              "notes": "See <a href='https://crbug.com/331608'>bug 331608</a>."
+              "version_added": false
             },
             "firefox": {
               "version_added": "1",
```

**Comment by foolip** [comment]
I'm trying to figure out how the Edge team could use this script after each Edge release to keep the data up-to-date, after a discussion with @Reezaali last week.

Unfortunately, the script is going to make incorrect changes based on things that were in IE but were removed in Edge 12. The natural way to avoid that would be `--source chrome`, but that doesn't work now, and `npm run mirror -- --source chrome -m bool edge` prints this warning:

> Warning: Edge does not support --source parameter.

@queengooborg that did work previously. Is there an alternative command that would be suitable for running for every release now?

**Comment by queengooborg** [comment]
I was going to suggest the `--target_version` parameter but looking through the code I realize that's not a thing yet... :D  I have a PR open, #6170, that adds that functionality in -- it's created specifically for the purpose of mirroring data for a new release for Samsung Internet, Opera, and now Edge!

**Comment by queengooborg** [comment]
Even if we might be considering removing Internet Explorer, I'd still like to get this landed as it provides important infrastructure changes to this script that I definitely expect us to use for other browsers as well.

**Comment by foolip** [comment]
@queengooborg I just tried `npm run mirror -- -m bool edge` again and it's making many changes that don't pass the lint. Two of them are:

```diff
--- a/api/Document.json
+++ b/api/Document.json
@@ -4946,9 +4946,7 @@
             "chrome_android": {
               "version_added": false
             },
-            "edge": {
-              "version_added": false
-            },
+            "edge": [],
             "firefox": {
               "version_added": false
             },

```

```diff
--- a/webextensions/manifest/manifest_version.json
+++ b/webextensions/manifest/manifest_version.json
@@ -33,9 +33,7 @@
                 "version_added": "4",
                 "version_removed": "21"
               },
-              "edge": {
-                "version_added": false
-              },
+              "edge": {},
               "firefox": {
                 "version_added": false
               },
```

Running `npm run mirror -- -m always edge` also makes changes that don't pass the lint, like this one:

```diff
--- a/svg/elements/discard.json
+++ b/svg/elements/discard.json
@@ -15,7 +15,7 @@
               "version_removed": "81"
             },
             "edge": {
-              "version_added": "79",
+              "version_added": "34",
               "version_removed": "81"
             },
             "firefox": {
```

Some manual fixing after running the script is expected, but the above seem like errors the script should be able to avoid.

**Comment by foolip** [comment]
So what should we do with Edge mirroring?

I think that with https://github.com/mdn/browser-compat-data/issues/15083 we will not want to supporting mirroring from both IE and Chrome, but just Chrome. I would really prefer to make that same change here, so that we can reuse more of the code.

**Comment by queengooborg** [comment]
> I think that with #15083 we will not want to supporting mirroring from both IE and Chrome, but just Chrome.

I opened up #16130 to change the default to Chrome, which should help with #15083!

**Comment by queengooborg** [comment]
I just fixed up the logic in regards to those three odd statements, and I have confirmed they don't negatively impact other statements as well!

**Comment by foolip** [comment]
I've tried `npm run mirror -- -m always edge` again, now it throws with an error.

It seems like the approach here is very brittle since there are new kinds of errors with each iteration. Can we simplify this to supporting mirroring from IE _or_ Chrome, not both at the same time? In other words, support `--source`, and have it default to Chrome.

Then it would be straightforward to review the changes that `npm run mirror -- --source ie edge` would make, and after that we could pretty much forget about that code path.

**Comment by queengooborg** [comment]
I just pushed a few changes that allow for selecting a source for Edge, and fixed the error I think you might've experienced!  I still have the dual browsers selected as the default for mirroring, but a `--source` argument is now allowed once again, though this time with a validation to ensure it's only one of three options we've scripted.

**Comment by queengooborg** [comment]
Made the default as just `chrome` now actually, with build-time mirroring coming up it's less likely that we'll want to mirror IE data.

**Comment by OrToR-KJM** [comment]
내 Galaxy에서 보냄
-------- 원본 이메일 --------발신: queengooborg ***@***.***> 날짜: 22/5/18  15:50  (GMT+09:00) 받은 사람: mdn/browser-compat-data ***@***.***> 참조: Subscribed ***@***.***> 제목: Re: [mdn/browser-compat-data] Mirroring script: Edge mirrors from both IE AND Chrome at the same time (#6582) 
Merged #6582 into main.
